lms/djangoapps/courseware/courses.py

```diff
@@ -118,7 +128,7 @@ def get_course_with_access(user, action, course_key, depth=0, check_if_enrolled=
       be plugged in as additional callback checks for different actions.
     """
     course = get_course_by_id(course_key, depth)
-    check_course_access(course, user, action, check_if_enrolled, check_survey_complete)
+    check_course_access_with_redirect(course, user, action, check_if_enrolled, check_survey_complete, check_if_authenticated)
```

Comment:
This is the crux of this refactoring - we added an extra layer between get_course_with_access and check_course_access which is responsible for the redirects that much of the existing code relies on.  We didn't want that layer - nor do we use this to get the course - so for the courseware_api endpoint we call check_course_access directly.